xds/internal/xdsclient/e2e_test/cds_watchers_test.go

```diff
@@ -0,0 +1,869 @@
+			updateCh1 := testutils.NewChannel()
+			cdsCancel1 := client.WatchCluster(test.resourceName, func(u xdsresource.ClusterUpdate, err error) {
+				updateCh1.Send(xdsresource.ClusterUpdateErrTuple{Update: u, Err: err})
+			})
+			defer cdsCancel1()
+			updateCh2 := testutils.NewChannel()
+			cdsCancel2 := client.WatchCluster(test.resourceName, func(u xdsresource.ClusterUpdate, err error) {
+				updateCh2.Send(xdsresource.ClusterUpdateErrTuple{Update: u, Err: err})
+			})
+
+			// Configure the management server to return a single cluster
+			// resource, corresponding to the one we registered watches for.
+			resources := e2e.UpdateOptions{
+				NodeID:         nodeID,
+				Clusters:       []*v3clusterpb.Cluster{test.watchedResource},
+				SkipValidation: true,
+			}
+			ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+			defer cancel()
+			if err := mgmtServer.Update(ctx, resources); err != nil {
+				t.Fatalf("Failed to update management server with resources: %v, err: %v", resources, err)
+			}
+
+			// Verify the contents of the received update.
+			if err := verifyClusterUpdate(ctx, updateCh1, test.wantUpdateV1); err != nil {
+				t.Fatal(err)
+			}
+			if err := verifyClusterUpdate(ctx, updateCh2, test.wantUpdateV1); err != nil {
+				t.Fatal(err)
+			}
+
+			// Cancel the second watch and force the management server to push a
+			// redundant update for the resource being watched. Neither of the
+			// two watch callbacks should be invoked.
+			cdsCancel2()
+			if err := mgmtServer.Update(ctx, resources); err != nil {
+				t.Fatalf("Failed to update management server with resources: %v, err: %v", resources, err)
+			}
+			if err := verifyNoClusterUpdate(ctx, updateCh1); err != nil {
+				t.Fatal(err)
+			}
+			if err := verifyNoClusterUpdate(ctx, updateCh2); err != nil {
+				t.Fatal(err)
+			}
+
+			// Update to the resource being watched. The un-cancelled callback
+			// should be invoked while the cancelled one should not be.
+			resources = e2e.UpdateOptions{
+				NodeID:         nodeID,
+				Clusters:       []*v3clusterpb.Cluster{test.updatedWatchedResource},
+				SkipValidation: true,
+			}
+			if err := mgmtServer.Update(ctx, resources); err != nil {
+				t.Fatalf("Failed to update management server with resources: %v, err: %v", resources, err)
+			}
+			if err := verifyClusterUpdate(ctx, updateCh1, test.wantUpdateV2); err != nil {
+				t.Fatal(err)
+			}
+			if err := verifyNoClusterUpdate(ctx, updateCh2); err != nil {
+				t.Fatal(err)
+			}
+		})
+	}
+}
+
+// TestCDSWatch_ThreeWatchesForDifferentResourceNames covers the case with three
+// watchers (two watchers for one resource, and the third watcher for another
+// resource), exist across two cluster resources.  The test verifies that an
+// update from the management server containing both resources results in the
+// invocation of all watch callbacks.
+//
+// The test is run with both old and new style names.
```

Comment:
I'm suggesting changing docstring specifying you use old and new in same resource, rather than a different test. Just to keep explanations consistent with docstrings earlier explaining you have seperate tests for old and new style resources. Here and the other place I commented about this explicit stating of how it uses old a new style names.

Reply:
Oh whoops lol on wrong partition I was thinking about.